lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -275,9 +283,126 @@ bool PatchEntryPointMutate::lowerCpsOps(Function *func) {
+  for (auto *ret : retInstrs) {
+    auto *vspTy = PointerType::get(func->getContext(), ADDR_SPACE_PRIVATE);
+    exitInfos.push_back(
+        CpsExitInfo(ret->getParent(), {builder.getInt32(0), PoisonValue::get(vspTy)}));
+    builder.SetInsertPoint(ret);
+    builder.CreateBr(tailBlock);
+    ret->eraseFromParent();
+  }
+
+  unsigned vgprNum = 0;
+  for (const auto &exit : exitInfos)
+    vgprNum = exit.vgpr.size() > vgprNum ? exit.vgpr.size() : vgprNum;
+
+  SmallVector<Value *> newVgpr;
+  builder.SetInsertPoint(tailBlock);
+
+  if (exitInfos.size() == 1) {
+    newVgpr.append(exitInfos[0].vgpr);
+  } else {
+    for (unsigned vgprIdx = 0; vgprIdx < vgprNum; vgprIdx++) {
+      // We always have the leading two fixed vgpr arguments: vcr, vsp. The other remaining payloads are i32 type.
+      Type *phiTy = vgprIdx < 2 ? exitInfos[0].vgpr[vgprIdx]->getType() : builder.getInt32Ty();
+      PHINode *phi = builder.CreatePHI(phiTy, exitInfos.size());
+      for (unsigned exitIdx = 0; exitIdx < exitInfos.size(); exitIdx++) {
+        if (vgprIdx < exitInfos[exitIdx].vgpr.size())
+          phi->addIncoming(exitInfos[exitIdx].vgpr[vgprIdx], exitInfos[exitIdx].pred);
+        else
+          phi->addIncoming(PoisonValue::get(builder.getInt32Ty()), exitInfos[exitIdx].pred);
+      }
+      newVgpr.push_back(phi);
+    }
+  }
+  // Packing VGPR arguments.
+  Value *vgprArg = mergeIntoStruct(builder, newVgpr);
+
+  // Packing SGPR arguments(user data) into vector of dwords.
+  SmallVector<Value *> userData;
+  for (unsigned idx = 0; idx != numUserdata; ++idx)
+    userData.push_back(func->getArg(idx));
+
+  SmallVector<Value *> userDataDwords;
+  splitIntoI32(layout, builder, userData, userDataDwords);
+  Value *userDataVec = mergeDwordsIntoVector(builder, userDataDwords);
+  //    tail:
+  //      Merge vgpr values from different exits.
+  //      Check if we have pending cps call
+  //      If no cps call, jump to return block.
+  //    chain:
+  //      Jump to next cps function.
+  //    ret:
+  //      ret void
+  unsigned waveSize = m_pipelineState->getShaderWaveSize(m_shaderStage);
+  Type *waveMaskTy = builder.getIntNTy(waveSize);
+  auto *chainBlock = BasicBlock::Create(func->getContext(), "chain.block", func);
+  auto *retBlock = BasicBlock::Create(func->getContext(), "ret.block", func);
+  auto *vcr = builder.CreateExtractValue(vgprArg, 0);
+  auto *vcrTy = vcr->getType();
+
+  Value *vcrWwm;
+  if (isCpsFunc) {
+    // FIXME: Switch to LLVM intrinsic
+    FunctionType *setInactiveChainTy = FunctionType::get(vcrTy, {vcrTy, vcrTy}, false);
+    auto setInactiveChain = Function::Create(setInactiveChainTy, GlobalValue::ExternalLinkage,
+                                             "llvm.amdgcn.setinactive.chain.arg", func->getParent());
+    vcrWwm = builder.CreateCall(setInactiveChain, {vcr, func->getArg(numUserdata)});
+  } else {
+    vcrWwm = builder.CreateIntrinsic(Intrinsic::amdgcn_set_inactive, vcrTy, {vcr, builder.getInt32(0)});
+  }
```

Comment:
The only case where we're not in a CPS function is the entry shader function. That one doesn't need any of the WWM logic.

Ideally, we would even recognize if `vcr` is uniform, but that's more difficult to pull off. We can postpone addressing that to a later time.

Reply:
This is used to handle the case that the entry point shader was launched with inactive threads in a wave. I think this is possible in Vulkan compute shader? In this case, the inactive lanes of vcr should be cleared to zero to make scheduler work correctly. Sounds reasonable?

Reply:
Yes, compute shaders can be launched with inactive threads in a wave, but overall this still wouldn't work:

* The entry point shader would jump/chain to the correct place
* However, `llvm.amdgcn.cs.chain` would still leave the values in inactive lanes of VGPRs unchanged. So `vcr` in the next CPS function would contain garbage in inactive lanes

I don't see an easy way to make this work, and it's not necessary in practice for what we want to do. And just not using any form of `set.inactive` should make the assembly code in the entry point shader a bit cheaper.

Reply:
>     * However, `llvm.amdgcn.cs.chain` would still leave the values in inactive lanes of VGPRs unchanged. So `vcr` in the next CPS function would contain garbage in inactive lanes

I get the point that `llvm.amdgcn.cs.chain` call will leave the inactive lanes unchanged, but how the vcr in the next cps function contain garbage if we clear lanes (by doing amdgcn.set.inactive) that were inactive at shader dispatch? All the lanes would have been set with meaningful data(either zero or target cps vcr).

Reply:
What `amdgcn.set.inactive` does conceptually only affects the SSA value that is defined by the `amdgcn.set.inactive`. It doesn't affect what's in the "physical" VGPR holding the vcr argument.

That's why the calculation in the entry point shader itself would be correct, but then the values are again undefined in the following CPS function.